1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,611 @@
+```
+
+Output:
+
+```output
+Reached
+```
+
+#### Example 8
+
+```powershell
+false || Write-Output "Command failed" && Write-Output "Backup"
+```
+
+Output:
+
+```output
+Command failed
+Backup
+```
+
+Also see: [intended test cases for implementation](https://github.com/PowerShell/PowerShell/blob/f8b899e42c86957a1b58273be0029f40a67bf1b6/test/powershell/Language/Operators/BashControlOperator.Tests.ps1).
+
+## Specification
+
+### Grammar
+
+Pipeline chains will be implemented using the already reserved `&&` and `||` operators.
+These operators will have the following grammar:
+
+```none
+statement:
+    | ... <existing statements except pipeline>   # Other statements
+    | variable_expression "=" statement           # Assignment
+    | pipeline_chain ["&"]                        # Pipeline chains
+
+pipeline_chain:
+    | pipeline
+    | pipeline_chain "&&" [newlines] pipeline
+    | pipeline_chain "||" [newlines] pipeline
+```
+
+#### Optional pipeline chaining
+
+A pipeline chain is a chain of one or more pipelines
+and takes the place of a pipeline in the current PowerShell syntax.
+
+In the degenerate case, a pipeline chain of a single pipeline
+will work exactly as pipelines currently do in PowerShell,
+so there is no breakage to the existing grammar.
+
+#### Line continuation
+
+After a pipeline chain operator,
+any newlines will be skipped in anticipation of the following pipeline.
+
+For example, the following would be a single pipeline chain:
+
+```powershell
+cmd1 &&
+    cmd2 ||
+    cmd3
+```
+
+If the end of file is reached after a pipeline chain operator,
+incomplete input will be reported so that integrating tools
+will know to keep prompting for input.
+
+#### Left-associativity
+
+Like in POSIX shells, pipeline chain operators will be left associative,
+meaning chain operators will group from left to right.
+
+For example, given the following pipeline chain:
+
+```powershell
+cmd1 && cmd2 || cmd3
+```
+
+This will be grouped as:
+
+```none
+[[cmd1 && cmd2] || cmd3]
+```
+
+As a syntax tree, this would look like:
+
+```none
+            "||"
+          /      \
+        "&&"     "cmd3"
+      /      \
+   "cmd1"    "cmd2"
+```
+
+With the syntax tree deepening on the left as more operators are chained.
+
+Semantically, this means that `cmd1 && cmd2` would be evaluated first,
+and its result used to govern the evaluation of `|| cmd3`.
+
+#### Higher precedence than `&` and `;`
+
+Pipeline chain operators will have higher precedence than
+pipeline background operators (`&`)
+or statement separators (`;`).
+
+For example:
+
+```none
+cmd1 && cmd2 &
+```
+
+Will bind as:
+
+```none
+[[cmd1 && cmd2] &]
+```
+
+Having the syntax tree:
+
+```none
+            <statement>
+           /           \
+         "&&"          "&"
+        /    \
+     "cmd1"  "cmd2"
+```
+
+The consequence of this will be that an entire pipeline chain
+can be sent to a background job for evaluation,
+rather than individual pipelines within it.

```

Comment:
Parentheses always set `$?` to true, so it will always succeed.

That might seem unhelpful, but I believe the solution there is to ensure `$?` is set properly.

Also, the `&` operator is translated into `Start-Job -ScriptBlock { ... }`. So `<expr> &` will succeed as long as `Start-Job` succeeds.